R/tm_g_response.R

```diff
@@ -234,7 +234,44 @@ srv_g_response <- function(id,
   moduleServer(id, function(input, output, session) {
     data_extract <- list(response = response, x = x, row_facet = row_facet, col_facet = col_facet)
 
-    selector_list <- teal.transform::data_extract_multiple_srv(data_extract, data)
+    rule_diff <- function(other) {
+      function(value) {
+        othervalue <- selector_list()[[other]]()$select
+        if (identical(value, othervalue))
+          "Row and column facetting variables must be different."
+      }
+    }
+
+    selector_list <- teal.transform::data_extract_multiple_srv(
+      data_extract = data_extract,
+      datasets = data,
+      select_validation_rule = list(
+        response = shinyvalidate::sv_required("Please define a column for the response variable"),
+        x = shinyvalidate::sv_required("Please define a column for X variable"),
+        row_facet = shinyvalidate::compose_rules(
+          shinyvalidate::sv_optional(),
+          ~ if (length(.) > 1) "There must be 1 or no row facetting variable.",
+          crule(rule_diff("col_facet"), !is.null(input$row_facet))
+        ),
+        col_facet = shinyvalidate::compose_rules(
+          shinyvalidate::sv_optional(),
+          ~ if (length(.) > 1) "There must be 1 or no column facetting variable.",
+          crule(rule_diff("row_facet"), !is.null(input$col_facet))
+        )
+      )
+    )
+
+    iv_r <- reactive({
+      iv_facet <- shinyvalidate::InputValidator$new()
+      iv_child <- teal.transform::compose_and_enable_validators(iv_facet, selector_list,
+                                                                validator_names = c("row_facet", "col_facet"))
+      iv_child$condition(~ isTRUE(any(c("row_facet", "col_facet") %in% names(selector_list()))))
+
+      iv <- shinyvalidate::InputValidator$new()
+      iv$add_rule("ggtheme", shinyvalidate::sv_required("Please select a theme"))
+      iv$add_validator(iv_child)
+      teal.transform::compose_and_enable_validators(iv, selector_list, validator_names = c("x", "response"))
```

Comment:
so is `compose_and_enable_validators` now smart enough to not need iv_child ?

Reply:
How do you mean? `iv_child` is now part of `iv` and validators from `selector_list` are added along.

Reply:
So could we just do 

```
iv_r <- reaactive({
  iv <- shinyvalidate::InputValidator$new()
  iv$add_rule("ggtheme", shinyvalidate::sv_required("Please select a theme"))
  teal.transform::compose_and_enable_validators(iv, selector_list)
})
```

As the new `intersect` in teal.transform::compose_and_enable_validators` will handle the `iv_child$condition` ?

Reply:
Ah, you are correct.